.github/workflows/self-test.yml

```diff
@@ -0,0 +1,20 @@
+name: Run the project's self-tests
+on: [push, pull_request]
```

Comment:
do we need `push` and `pull_request`? i think `push` covers all commits so pull request would be unnecessary here. or we could restrict push events to only the main branch and let the pull request trigger catch everything else.